lib/scnnr/client.rb

```diff
@@ -56,6 +54,20 @@ def post_connection(uri, options = {})
       Connection.new(uri, :post, options[:api_key], options[:logger])
     end
 
+    def request_create(path, options)
+      options, fetch_timeout = restrict_timeout(merge_options(options))
```

Comment:
I think your approach is perfect 👍  I confirmed that the library behaves as expected.

And then this is my opinion about code design, I've added `PollingManager` class to handle timeouts, so I think the operation of `restrict_timeout` should be handled by `PollingManager` as well (maybe `TimeoutError` handling is also the same).

I've been thinking like following code (this is sample, may be not working).

```
module Scnnr
  class PollingManager
    class << self
      def start(client, options, &block)
        timeout = options.delete(:timeout)
        used_timeout = [timeout, MAX_TIMEOUT].min
        recognition = block.call(options.merge(timeout: used_timeout))
        t = timeout - used_timeout
        self.new(t).polling(client, recognition.id, options) if recognition.queued? && t >= 0
      end
    end
  end
end

module Scnnr
  class Client
    def recognize_url(url, options = {})
      options = merge_options(options)
      PollingManager.start(self, options) do |opts|
        uri = construct_uri('remote/recognitions', opts)
        response = post_connection(uri, opts).send_json({ url: url })
        break handle_response(response, opts)
      end
    end
  end
end
```
Let me know what you think :)

Reply:
I see what you mean, and I was thinking along the same lines. My approach was to make `PollingManager` even simpler so it just handles the timeouts, and pass everything in when it's used, e.g.

```ruby
    def recognize_url(url, options = {})
      PollingManager.new(options).
        stop_when { |recognition| recognition.finished? }.
        once { |options| request_recognize_url(url, options) }.
        repeat { |options, last_recognition| fetch(last_recognition, options) }
    end
```

The downside of this is that `Client` ends up doing a lot of stuff and becomes quite complex.

I like your idea, but it ends up being:

```ruby
    def self.start(client, options)
      total_timeout = options.delete(:timeout)
      used_timeout = [total_timeout, MAX_TIMEOUT].min

      recognition = yield options.merge(timeout: used_timeout)
      if recognition.queued? && (total_timeout - used_timeout).positive?
        new(total_timeout - used_timeout).polling(client, recognition.id, options)
      else
        recognition
      end
    end
```

So the `start` method may or may not actually use a `PollingManager`, which doesn't quite feel right ... feel free to disagree :)

Maybe making `PollingManager` do nothing if the timeout is zero, and changing `fetch` to be like this would work?

```
    def fetch(recognition_id, options = {})
      return request(recognition_id, options) if options.delete(:polling) == false

      options = merge_options(options)
      PollingManager.start(self, options) do |opts|
        request(recognition_id, opts)
      end
    end
```

Sorry this got so long! Any ideas/opinions?

Reply:
Thanks for your reply! It is a lot of fun to discuss with you :)

Why do you think `PollingManager.start` is not good? Is it solved by splitting `PollingManager.start` into multiple methods? Or this approach itself is not good?  
If `Client` handles timeouts directly, I think it is more complicated... 

According to Object Oriented Design, I think responsibility of each of the classes as following.

- Client
  - This class should work as a hub for the others. This should not handle timeouts directly, should not request to API directly, should not handle responses directly. This just use the others.
- Connection 
  - This class should request to API.
- PollingManager
  - This class should handle timeouts, should decide wheter to keep polling, should handle timeout errors (sorry there is a bug in current master branch) .
- Response
  - This class should check response status codes, should decide whether to return an error.

For example in `Client#restrict_timeout`, `Client` handles `PollingManager::MAX_TIMEOUT` directly. It seems like this is outside `Client`'s area of responsibility. Even if this is correct, `MAX_TIMEOUT` constant should be defined in `Client` class.

For your reference, I've published my sample implementation [here](https://github.com/NEWROPE/scnnr-ruby/compare/master...danimal141:feature/timeout-handling). (sorry I've implemented in a hurry, so maybe tests are failed yet...)
maybe, working as following.

```
# Setup ENDPOINT_BASE = 'http://localhost:8000'
cli = Scnnr::Client.new
rec = cli.recognize_url 'some url', timeout:150

I, [2017-10-25T12:05:46.095212 #13375]  INFO -- : Started POST http://localhost:8000/v1/remote/recognitions?timeout=25
I, [2017-10-25T12:05:46.104424 #13375]  INFO -- : Started GET http://localhost:8000/v1/recognitions/20170829/ed4c674c-7970-4e9c-9b26-1b6076b36b49?timeout=25
I, [2017-10-25T12:05:46.110301 #13375]  INFO -- : Started GET http://localhost:8000/v1/recognitions/20170829/ed4c674c-7970-4e9c-9b26-1b6076b36b49?timeout=25
I, [2017-10-25T12:05:46.111826 #13375]  INFO -- : Started GET http://localhost:8000/v1/recognitions/20170829/ed4c674c-7970-4e9c-9b26-1b6076b36b49?timeout=25
I, [2017-10-25T12:05:46.113565 #13375]  INFO -- : Started GET http://localhost:8000/v1/recognitions/20170829/ed4c674c-7970-4e9c-9b26-1b6076b36b49?timeout=25
I, [2017-10-25T12:05:46.115192 #13375]  INFO -- : Started GET http://localhost:8000/v1/recognitions/20170829/ed4c674c-7970-4e9c-9b26-1b6076b36b49?timeout=25
Scnnr::TimeoutError: recognition timed out
```

Let me know your opinion!

Reply:
Thanks for the breakdown, I see where you're coming from now.

I guess having done a lot of functional and reactive programming recently, I would think about this problem in a somewhat different way - if I was doing this from scratch I would move all the logic into single-responsibility command objects (or service objects, or use cases, or actions) like FetchRecognition and CreateRecognition, and have each command handle its own timeout behaviour (still with some helper classes and things to avoid duplication). So I think my unease about `PollingManager` is because it becomes less single-responsibility than I would like (e.g. you can start polling in two ways using `start` or `new`, then it both handles timeouts and does logic on the client and recognition). I don't necessarily consider it a problem, just a different way of thinking about things, or perhaps different priorities than what I'm used to.

It's hard to explain in a simple way, so sorry if it doesn't make sense! In conclusion, your approach makes sense to me now, and I can't think of an alternative that's obviously better. However, I wonder if a clearer path to making `PollingManager` and `Client` more modular and less interconnected might appear with future changes.

By the way, I really like your solution to timeout errors, and the removal of `async` from `Response` - that caused me some confusion while I was working on this.

I'm happy to fix this implementation and get specs passing again, but I think it would end up basically like yours now! Depends whether you're more interested in the code or the discussion :)

Reply:
Oh, I see! I was thinking your approach is kind of functional. I know your feelings because I have experience of reactive programming in iOS development, but I don't do it in Ruby development...
I want to see your implementation from scratch if there's any chance. I can expect pretty good stuff to be completed :)

> By the way, I really like your solution to timeout errors, and the removal of async from Response - that caused me some confusion while I was working on this.

Sorry, `async` in `Response` is so confusing... it existed only to handle `TimeoutError`, but `Response` should not do it anyways...

> I'm happy to fix this implementation and get specs passing again, but I think it would end up basically like yours now! Depends whether you're more interested in the code or the discussion :)

Really? So helpful! I would like to ask you to do that work! please fix my rough implementation.

Reply:
> please fix my rough implementation.

Done - and added/moved a few specs for the new functionality. See what you think :)

> I want to see your implementation from scratch if there's any chance.

I'll try and do that over the next few days. I'm really interested to try, but it's hard to find time at the moment!

Reply:
Thanks, I've checked the behaviour and LGTM!

> I'll try and do that over the next few days. I'm really interested to try, but it's hard to find time at the moment!

Oh, thanks. But please don't push yourself so hard :)